harness/tests/storage/util.py

```diff
@@ -186,3 +262,213 @@ def timed_backoff_sync_check(expected_files: Dict[str, bytes]) -> None:
+
+    # 3.test downloading with and w/o selector
+    download_dir = tmp_path.joinpath(f"test3_download_{pex.distributed.rank}")
+    if pex.distributed.rank == 0:
+
+        def selector(x: str) -> bool:
+            return x in EXPECTED_FILES_N0
+
+    else:
+        selector = None
+
+    try:
+        checkpoint_context.download(storage_id, download_dir, selector=selector)
+        if pex.distributed.rank == 0:
+            validate_checkpoint(download_dir, expected_files=EXPECTED_FILES_N0)
+        else:
+            assert not download_dir.exists()
+    finally:
+        shutil.rmtree(download_dir, ignore_errors=True)
+
+    # cleanup
+    if pex.distributed.rank == 0 and clean_up is not None:
+        clean_up(storage_id, storage_manager)
+    logging.info(f"before broadcast {pex.distributed.rank}")
+    pex.distributed.allgather(None)
+    logging.info(f" after broadcast {pex.distributed.rank}")
+
+    # 1. upload sharded data from rank 0 only
+    storage_id = checkpoint_context.upload(
+        ckpt_dir if pex.distributed.rank == 0 else None, metadata, shard=True
+    )
+    download_dir = tmp_path.joinpath(f"test4_download_{pex.distributed.rank}")
+    checkpoint_context.download(storage_id, download_dir)
+    validate_checkpoint(download_dir, expected_files=EXPECTED_FILES_N0)
+
+    if pex.distributed.rank == 0 and clean_up is not None:
+        clean_up(storage_id, storage_manager)
+    pex.distributed.broadcast(None)
+
+    # 2. upload sharded data from rank 1 only
+    # metadata should be saved and uploaded as well
+    storage_id = checkpoint_context.upload(
+        ckpt_dir if pex.distributed.rank == 1 else None, metadata, shard=True
+    )
+    download_dir = tmp_path.joinpath(f"test5_download_{pex.distributed.rank}")
+    checkpoint_context.download(storage_id, download_dir)
+    validate_checkpoint(
+        download_dir,
+        expected_files={**EXPECTED_FILES_N1, **{"metadata.json": '{\n  "steps_completed": 1\n}'}},
+    )
+
+    if pex.distributed.rank == 0 and clean_up is not None:
+        clean_up(storage_id, storage_manager)
+    pex.distributed.broadcast(None)
+
+
+def run_storage_store_restore_sharded_test(
+    pex: parallel.Execution,
+    storage_manager: storage.StorageManager,
+    clean_up: Optional[Callable] = None,
+) -> None:
+    # create checkpoint context
+    checkpoint_context = create_checkpoint_context(pex, storage_manager)
+
+    # upload sharded data
+    metadata: Optional[Dict[str, int]] = None
+    if pex.distributed.rank == 0:
+        metadata = {"steps_completed": 1}
+
+    with checkpoint_context.store_path(metadata, shard=True) as (path, storage_id):
+        logging.info(f"storage_id={storage_id}")
+        # create "local" file structure
+        if pex.distributed.rank == 0:
+            create_checkpoint(path, EXPECTED_FILES_N0)
+        else:
+            create_checkpoint(path, EXPECTED_FILES_N1)
+
+    pex.distributed.broadcast(None)
+    selector: Optional[Callable[[str], bool]]
```

Comment:
question: if you put just a bare `selector = None` here, can you drop the `else: selector=None` clause, and will mypy be ok with the resulting code?

I know there's some incantation where mypy will automatically figure out that something is an Optional[Thing], but I can't quite remember what it is.

Reply:
Removing `else: selector=None` would make rank=1 to re-use the selector created in the above test. I can change the order of the tests to avoid this problem, but I think it is better to be explicit about what a selector is for a given test than to rely on ordering.